src/python/pants/engine/native.py

```diff
@@ -257,33 +257,58 @@ def get_build_cflags():
 
 def bootstrap_c_source(output_dir, module_name=NATIVE_ENGINE_MODULE):
   """Bootstrap an external CFFI C source file."""
+
+  tempdir = safe_mkdtemp()
+
   safe_mkdir(output_dir)
 
-  output_prefix = os.path.join(output_dir, module_name)
-  c_file = '{}.c'.format(output_prefix)
-  env_script = '{}.cflags'.format(output_prefix)
+  temp_output_prefix = os.path.join(tempdir, module_name)
+  real_output_prefix = os.path.join(output_dir, module_name)
+  temp_c_file = '{}.c'.format(temp_output_prefix)
+  c_file = '{}.c'.format(real_output_prefix)
+  env_script = '{}.cflags'.format(real_output_prefix)
 
   ffibuilder = cffi.FFI()
   ffibuilder.cdef(CFFI_TYPEDEFS)
   ffibuilder.cdef(CFFI_HEADERS)
   ffibuilder.cdef(CFFI_EXTERNS)
   ffibuilder.set_source(module_name, CFFI_TYPEDEFS + CFFI_HEADERS)
-  ffibuilder.emit_c_code(six.binary_type(c_file))
+  ffibuilder.emit_c_code(six.binary_type(temp_c_file))
+
+  # Work around https://github.com/rust-lang/rust/issues/36342 by renaming initnative_engine to
+  # wrapped_initnative_engine so that the rust code can define the symbol initnative_engine.
+  #
+  # If we dont do this, we end up at the mercy of the implementation details of rust's stripping
+  # and LTO. In the past we have found ways to trick it into not stripping symbols which was handy
+  # (it kept the binary working) but inconvenient (it was relying on unspecified behavior, it meant
+  # our binaries couldn't be stripped which inflated them by 2~3x, and it reduced the amount of LTO
+  # we could use, which led to unmeasured performance hits).
+  with open(temp_c_file, 'rb') as f:
+    lines = f.readlines()
+  for i, line in enumerate(lines):
+    # This is an optimistic heuristic for renaming the initnative_engine symbol and nothing else.
+    if line.startswith(b'init{}'.format(module_name)):
+      lines[i] = 'wrapped_' + line
+  file_content = ''.join(lines)
```

Comment:
how about using a generator inline here to mutate on the fly? e.g.

```
def transform_file(f):
  with open(f, 'rb') as fh:
    for line in fh:
      if line.startswith(b'init{}'.format(module_name)):
        yield 'wrapped_' + line
      else:
        yield line

file_content = ''.join(transform_file(temp_c_file))
```

Reply:
Done :)